match/nn.py

```diff
@@ -0,0 +1,101 @@
+"""
+~SAM~
+COPIED DIRECTLY FROM MATCH LIBRARY FOR REVERSE ENGINEERING PURPOSES
```

Comment:
I won't review this file just yet. We'll come back to it when the linear algebra is working.